File: code/PBTDemos/tests/HelloWorldFsharp.Tests/FizzBuzzerTests.fs
module FizzBuzzerTests

open System
open Xunit
open FsCheck
open FsCheck.FSharp
open FsCheck.Xunit

(*
FizzBuzz is actually an "advanced" example in the PBT context because it requires knowledge of
Arbitraries and Generators
*)
module FizzBuzzing =

  let fizzbuzz n =
    match n % 3, n % 5 with
    | 0, 0 -> "FizzBuzz"
    | 0, _ -> "Fizz"
    | _, 0 -> "Buzz"
    | _ -> n.ToString()

  module FactArbPropCheckPattern =

    [<Fact>]
    let ``Is a number`` () =
      let arb =
        Gen.choose (0, 100)
        |> Gen.filter (fun x -> x % 3 <> 0 && x % 5 <> 0)
        |> Arb.fromGen

      let property number = (fizzbuzz number) = number.ToString()
      Prop.forAll arb property |> Check.QuickThrowOnFailure

    [<Fact>]
    let ``Divisible by 3`` () =
      let arb = Gen.choose (0, 100) |> Gen.map (fun x -> x * 3) |> Arb.fromGen
      let property number = (fizzbuzz number).StartsWith("Fizz")
      Prop.forAll arb property |> Check.QuickThrowOnFailure

    [<Fact>]
    let ``Divisible by 5`` () =
      let arb =
        Gen.choose (0, 100)
        |> Gen.map (fun x -> x * 5)
        |> Gen.filter (fun x -> x % 3 <> 0)
        |> Arb.fromGen

      let property number = (fizzbuzz number).StartsWith("Buzz")
      Prop.forAll arb property |> Check.QuickThrowOnFailure

    [<Fact>]
    let ``Divisible by 15`` () =
      let arb = Gen.choose (0, 100) |> Gen.map (fun x -> x * 15) |> Arb.fromGen

      let property number =
        (fizzbuzz number).StartsWith("FizzBuzz")

      Prop.forAll arb property |> Check.QuickThrowOnFailure

  module XunitPropertyAttributePattern =

    let nonDivisibleNumberArb =
      Gen.choose (0, 100)
      |> Gen.filter (fun x -> x % 3 <> 0 && x % 5 <> 0)
      |> Arb.fromGen

    type NonDivisibleNumberArb =
      static member NonDivisibleNumber() = nonDivisibleNumberArb

    [<Property(Arbitrary = [| typeof<NonDivisibleNumberArb> |])>]
    let ``Non divisible numbers`` number = (fizzbuzz number) = number.ToString()

    let divisibleByThreeArb =
      Gen.choose (0, 100)
      |> Gen.map (fun x -> x * 3)
      |> Gen.filter (fun x -> x % 5 <> 0)
      |> Arb.fromGen

    type DivisibleByThreeArb =
      static member DivisibleByThree() = divisibleByThreeArb

    [<Property(Arbitrary = [| typeof<DivisibleByThreeArb> |])>]
    let ``Divisible by 3`` number = (fizzbuzz number) = "Fizz"

    let divisibleByFiveArb =
      Gen.choose (0, 100)
      |> Gen.map (fun x -> x * 5)
      |> Gen.filter (fun x -> x % 3 <> 0)
      |> Arb.fromGen

    type DivisibleByFiveArb =
      static member DivisibleByFive() = divisibleByFiveArb

    [<Property(Arbitrary = [| typeof<DivisibleByFiveArb> |])>]
    let ``Divisible by 5`` number = (fizzbuzz number) = "Buzz"

    let divisibleByFifteenArb =
      Gen.choose (0, 100) |> Gen.map (fun x -> x * 15) |> Arb.fromGen

    type DivisibleByFifteenArb =
      static member DivisibleByFifteen() = divisibleByFifteenArb

    [<Property(Arbitrary = [| typeof<DivisibleByFifteenArb> |])>]
    let ``Divisible by 15`` number = (fizzbuzz number) = "FizzBuzz"

(*
The above examples have one major flaw: The generation of test data is closely coupled to the implementation.

Here is a cleaner solution.

Finding these properties is the hard part...

Ported from Mark Seemann's Haskell examples:

https://web.archive.org/web/20240910144730/https://blog.ploeh.dk/2021/06/28/property-based-testing-is-not-the-same-as-partition-testing/
*)
module FizzBuzzingDoneCorrectly =

  let fizzBuzz n =
    match n % 3, n % 5 with
    | 0, 0 -> "FizzBuzz"
    | 0, _ -> "Fizz"
    | _, 0 -> "Buzz"
    | _, _ -> n.ToString()

  let fizzBuzzDeveloperFromHell1 n =
    if n % 15 = 0 then "FizzBuzz" else "2112"

  let fizzBuzzDeveloperFromHell2 n =
    match n % 3, n % 5 with
    | _, 0 -> "FizzBuzz"
    | 0, _ -> "Fizz"
    | _, _ -> "2112"

  let fizzBuzzDeveloperFromHell3 n =
    match n % 3, n % 5 with
    | 0, 0 -> "FizzBuzz"
    | 0, _ -> "Fizz"
    | _, 0 -> "Buzz"
    | _, _ -> "2112"

  [<Property>]
  let ``at least one number in 3 consecutive values`` (i: int) =
    let range = [ i .. i + 2 ]

    let tryToNumber (s: string) =
      match Int32.TryParse(s) with
      | true, value -> Some value
      | false, _ -> None

    let actual =
      range
      // |> List.map fizzBuzzDeveloperFromHell1
      // |> List.map fizzBuzzDeveloperFromHell2
      // |> List.map fizzBuzzDeveloperFromHell3
      |> List.map fizzBuzz
      |> List.map tryToNumber
      |> List.choose id
      |> List.length

    actual >= 1

  [<Property>]
  let ``only one Buzz in 5 consecutive values`` (i: int) =
    let range = [ i .. i + 4 ]

    let actual =
      range
      // |> List.map fizzBuzzDeveloperFromHell2
      // |> List.map fizzBuzzDeveloperFromHell3
      |> List.map fizzBuzz
      |> List.filter _.EndsWith("Buzz")
      |> List.length

    actual = 1

  [<Property>]
  let ``at least one literal Buzz in 10 values`` (i: int) =
    let range = [ i .. i + 9 ]

    let actual =
      range
      // |> List.map fizzBuzzDeveloperFromHell2
      // |> List.map fizzBuzzDeveloperFromHell3
      |> List.map fizzBuzz
      |> List.filter (fun s -> s = "Buzz")
      |> List.length

    actual >= 1

  [<Property>]
  let ``numbers round-trip (there-and-back-again)`` (i: int) =
    let range = [ i .. i + 2 ]

    let tryToNumber (s: string) =
      match Int32.TryParse(s) with
      | true, value -> Some value
      | false, _ -> None

    let actual =
      range
      |> List.map fizzBuzz
      |> List.map tryToNumber
      |> List.choose id
      |> List.forall (fun x -> List.contains x range)

    actual = true

File: code/PBTDemos/tests/HelloWorld.Tests/FizzBuzzerTests.cs
using FsCheck;
using FsCheck.Fluent;
using FsCheck.Xunit;

using JetBrains.Annotations;

namespace HelloWorld.Tests;

public static class FizzBuzzing
{
  public static string FizzBuzz(int n)
  {
    if (n % 3 == 0 && n % 5 == 0)
    {
      return "FizzBuzz";
    }

    if (n % 3 == 0)
    {
      return "Fizz";
    }

    if (n % 5 == 0)
    {
      return "Buzz";
    }

    return n.ToString();
  }
}

[UsedImplicitly]
public class FizzBuzzTests
{
  public class NaiveTests
  {
    [Fact]
    public void IsANumber()
    {
      var arb = Arb.From(
        Gen.Choose(0, 100)
          .Where(x => x % 3 != 0 && x % 5 != 0));

      Prop.ForAll(arb, number =>
          FizzBuzzing.FizzBuzz(number) == number.ToString())
        .QuickCheckThrowOnFailure();
    }

    [Fact]
    public void DivisibleByThree()
    {
      var arb = Arb.From(
        Gen.Choose(0, 100)
          .Select(x => x * 3));

      Prop.ForAll(arb, number =>
          FizzBuzzing.FizzBuzz(number).StartsWith("Fizz"))
        .QuickCheckThrowOnFailure();
    }

    [Fact]
    public void DivisibleByFive()
    {
      var arb = Arb.From(
        Gen.Choose(0, 100)
          .Select(x => x * 5)
          .Where(x => x % 3 != 0));

      Prop.ForAll(arb, number =>
          FizzBuzzing.FizzBuzz(number) == "Buzz")
        .QuickCheckThrowOnFailure();
    }

    [Fact]
    public void DivisibleByFifteen()
    {
      var arb = Arb.From(
        Gen.Choose(0, 100)
          .Select(x => x * 15));

      Prop.ForAll(arb, number =>
          FizzBuzzing.FizzBuzz(number) == "FizzBuzz")
        .QuickCheckThrowOnFailure();
    }
  }

  public class XunitPropertyAttributePattern
  {
    [Property(Arbitrary = [typeof(NonDivisibleNumberArbitrary)])]
    public void NonDivisibleNumbers(int number)
    {
      Assert.Equal(number.ToString(), FizzBuzzing.FizzBuzz(number));
    }

    [Property(Arbitrary = [typeof(DivisibleByThreeArbitrary)])]
    public void DivisibleByThree(int number)
    {
      Assert.Equal("Fizz", FizzBuzzing.FizzBuzz(number));
    }

    [Property(Arbitrary = [typeof(DivisibleByFiveArbitrary)])]
    public void DivisibleByFive(int number)
    {
      Assert.Equal("Buzz", FizzBuzzing.FizzBuzz(number));
    }

    [Property(Arbitrary = [typeof(DivisibleByFifteenArbitrary)])]
    public void DivisibleByFifteen(int number)
    {
      Assert.Equal("FizzBuzz", FizzBuzzing.FizzBuzz(number));
    }

    private class NonDivisibleNumberArbitrary
    {
      public static Arbitrary<int> NonDivisibleNumber()
      {
        return Arb.From(
          Gen.Choose(0, 100)
            .Where(x => x % 3 != 0 && x % 5 != 0));
      }
    }

    private class DivisibleByThreeArbitrary
    {
      public static Arbitrary<int> DivisibleByThree()
      {
        return Arb.From(
          Gen.Choose(0, 100)
            .Select(x => x * 3)
            .Where(x => x % 5 != 0));
      }
    }

    private class DivisibleByFiveArbitrary
    {
      public static Arbitrary<int> DivisibleByFive()
      {
        return Arb.From(
          Gen.Choose(0, 100)
            .Select(x => x * 5)
            .Where(x => x % 3 != 0));
      }
    }

    private class DivisibleByFifteenArbitrary
    {
      public static Arbitrary<int> DivisibleByFifteen()
      {
        return Arb.From(
          Gen.Choose(0, 100)
            .Select(x => x * 15));
      }
    }
  }

  public class PropertyTests
  {
    [Property]
    public void AtLeastOneNumberInThreeConsecutiveValues(int i)
    {
      var range = Enumerable.Range(i, 3);
      var actual = range
        .Select(FizzBuzzing.FizzBuzz)
        .Select(s => int.TryParse(s, out var num) ? (int?)num : null)
        .Count(x => x.HasValue);

      Assert.True(actual >= 1);
    }

    [Property]
    public void OnlyOneBuzzInFiveConsecutiveValues(int i)
    {
      var range = Enumerable.Range(i, 5);
      var actual = range
        .Select(FizzBuzzing.FizzBuzz)
        .Count(s => s.EndsWith("Buzz"));

      Assert.Equal(1, actual);
    }

    [Property]
    public void AtLeastOneLiteralBuzzInTenValues(int i)
    {
      var range = Enumerable.Range(i, 10);
      var actual = range
        .Select(FizzBuzzing.FizzBuzz)
        .Count(s => s == "Buzz");

      Assert.True(actual >= 1);
    }

    [Property]
    public void NumbersRoundTrip(int i)
    {
      var range = Enumerable.Range(i, 3).ToList();
      var actual = range
        .Select(FizzBuzzing.FizzBuzz)
        .Select(s => int.TryParse(s, out var num) ? (int?)num : null)
        .Where(x => x.HasValue)
        .All(x => range.Contains(x.Value));

      Assert.True(actual);
    }
  }
}